app/app.go

```diff
@@ -0,0 +1,105 @@
+package app
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/app"
+	"github.com/iov-one/weave/coin"
+	"github.com/iov-one/weave/orm"
+	"github.com/iov-one/weave/store/iavl"
+	"github.com/iov-one/weave/x"
+	"github.com/iov-one/weave/x/msgfee"
+	"github.com/iov-one/weave/x/multisig"
+	"github.com/iov-one/weave/x/sigs"
+	"github.com/iov-one/weave/x/utils"
+)
+
+// Authenticator returns authentication with multisigs
+// and public key signatues
+func Authenticator() x.Authenticator {
+	return x.ChainAuth(sigs.Authenticate{}, multisig.Authenticate{})
+}
+
+// Chain returns a chain of decorators, to handle authentication,
+// fees, logging, and recovery
+func Chain(authFn x.Authenticator, minFee coin.Coin) app.Decorators {
+
+	// TODO implement orderbook controller
+	return app.ChainDecorators(
+		utils.NewLogging(),
+		utils.NewRecovery(),
+		utils.NewKeyTagger(),
+		utils.NewSavepoint().OnCheck(),
+		sigs.NewDecorator(),
+		multisig.NewDecorator(authFn),
+		msgfee.NewFeeDecorator(),
```

Comment:
You probably want to add a `utils.NewSavepoint().OnDeliver()` before the fee decorator.

Reply:
Aaah I see now, these decorators always in the given order. So `utils.NewSavepoint().OnCheck()` sets a save point before sigs, and `utils.NewSavepoint().OnDeliver()` sets a savepoint thus enables msgfee to be taken even tx fails? Did I get all right?

Reply:
Actually, when you set `utils.NewSavepoint().OnDeliver()` before msg fee - msg fee will be discarded on failure.